pyro/ops/einsum/shared.py

```diff
@@ -0,0 +1,147 @@
+from __future__ import absolute_import, division, print_function
+
+import contextlib
+import numbers
+from collections import OrderedDict
+
+import opt_einsum
+
+_SHARING_STACK = []
+_CURRENT_BACKEND = []
+LAST_CACHE_SIZE = [0]  # for profiling
+
+
+@contextlib.contextmanager
+def shared_intermediates(cache=None):
+    """
+    Context in which :func:`~pyro.ops.einsum.contract` intermediate results are shared.
+    Note that intermediate computations will not be garbage collected until
+    1. this context exits, and
+    2. the yielded cache is garbage collected (if it was captured).
+
+    :param dict cache: an optional dict
+    :yields: a cache dict suitable for saving and reusing as an input
+    """
+    if cache is None:
+        cache = {}
+    _SHARING_STACK.append(cache)
```

Comment:
Will this correctly handle multiple backends - say, if we use this as a context manager inside of which we are using two different backends?

Reply:
Good point, I'll key the cache by backend as well.